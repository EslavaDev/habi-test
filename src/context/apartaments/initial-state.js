export const InitialStateApartaments = {
  apartaments: null,
  apartamentsMap: {},
  apartamentsLike: {},
  dateFilter: null,
  locationFilter: null,
};
