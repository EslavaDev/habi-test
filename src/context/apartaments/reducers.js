export const reducersApartaments = (state, action) => {
  switch (action.type) {
    case '@@FETCH':
      return {
        ...state,
        apartaments: action.payload.array,
        apartamentsMap: action.payload.json,
      };
    case '@@ADD': {
      return {
        ...state,
        ...action.payload,
      };
    }
    case '@@FILTER':
      return {
        ...state,
        apartamentsMap: action.payload,
      };
    case '@@REMOVE':
      return {
        ...state,
        apartamentsMap: action.payload.aparments,
        apartamentsLike: action.payload.likes,
      };
    case '@@LIKE':
      return {
        ...state,
        apartamentsLike: action.payload,
      };
    default:
      return state;
  }
};
