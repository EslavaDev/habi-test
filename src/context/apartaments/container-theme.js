import { useReducer } from 'react';
import { reducersApartaments } from './reducers';
import MockData from '../../mock/apartaments-mock.json';
import { convertArrayToJson, generateObjectWithCount } from '../../utils/converts';

export const useContainerApartaments = initialState => {
  let initial = initialState;

  const stateLocal = localStorage.getItem('apartments');
  if (stateLocal) {
    initial = JSON.parse(stateLocal);
  }
  const [state, dispatch] = useReducer(reducersApartaments, initial);
  const generateApartment = generateObjectWithCount(5);
  const generateOwner = generateObjectWithCount(5);

  const fetchApartaments = (refresh = false) => {
    let jsonMock = null;
    let arrayMock = null;
    if (refresh) {
      const localData = JSON.parse(stateLocal);
      console.log(localData);
      if (localData?.apartaments?.length > 0) {
        jsonMock = convertArrayToJson(localData.apartaments);
        arrayMock = localData.apartaments;
      } else {
        jsonMock = convertArrayToJson(MockData);
        arrayMock = MockData;
      }
    } else {
      const { apartaments } = state;
      jsonMock = convertArrayToJson(apartaments);
      arrayMock = apartaments;
    }
    localStorage.setItem('apartments', JSON.stringify({ ...state, apartaments: arrayMock, apartamentsMap: jsonMock }));
    dispatch({
      type: '@@FETCH',
      payload: {
        json: jsonMock,
        array: arrayMock,
      },
    });
  };

  const likeApartament = id => {
    const { apartamentsMap, apartamentsLike } = state;
    const newApartamentLike = { ...apartamentsLike };
    if (!newApartamentLike[id]) {
      newApartamentLike[id] = apartamentsMap[id];
    } else {
      delete newApartamentLike[id];
    }
    localStorage.setItem('apartments', JSON.stringify({ ...state, apartamentsLike: newApartamentLike }));
    dispatch({
      type: '@@LIKE',
      payload: newApartamentLike,
    });
  };

  const filterApartaments = ({ location, date }) => {
    const { apartaments } = state;
    const apartamentsFilter = apartaments.filter(apartment => {
      if (location && date) {
        const filter = apartment.location.includes(location) && apartment.date.includes(date);
        return filter;
      }
      if (location) {
        const filter = apartment.location.includes(location);
        return filter;
      }
      const filter = apartment.date.includes(date);
      return filter;
    });
    dispatch({
      type: '@@FILTER',
      payload: convertArrayToJson(apartamentsFilter),
    });
  };

  const removeApartment = id => {
    const { apartamentsMap, apartamentsLike } = state;
    const newApartmentMap = { ...apartamentsMap };
    const newApartmentsLike = { ...apartamentsLike };
    delete newApartmentMap[id];
    if (newApartmentsLike[id]) {
      delete newApartmentsLike[id];
    }
    const payload = {
      apartaments: newApartmentMap,
      apartamentsLike: newApartmentsLike,
    };
    localStorage.setItem('apartments', JSON.stringify({ ...state, ...payload }));
    dispatch({
      type: '@@REMOVE',
      payload: {
        aparments: newApartmentMap,
        likes: newApartmentsLike,
      },
    });
  };

  const addApartament = ({ owner: ownerTemp, ...values }) => {
    const { apartaments: apartamentsTemp, apartamentsMap: apartamentsMapTemp } = state;
    const ownerId = generateOwner.next().value;
    const apartamentId = generateApartment.next().value;

    const newValues = { ...values, id: `apt${apartamentId}`, owner: { ...ownerTemp, id: `own${ownerId}` } };

    const apartaments = [...apartamentsTemp, newValues];
    const apartamentsMap = { ...apartamentsMapTemp, [newValues.id]: newValues };

    const payload = {
      apartaments,
      apartamentsMap,
    };

    localStorage.setItem('apartments', JSON.stringify({ ...state, ...payload }));

    dispatch({
      type: '@@ADD',
      payload: {
        apartaments,
        apartamentsMap,
      },
    });
  };
  return {
    data: {
      ...state,
    },
    mutations: {
      fetchApartaments,
      likeApartament,
      filterApartaments,
      removeApartment,
      addApartament,
    },
  };
};
