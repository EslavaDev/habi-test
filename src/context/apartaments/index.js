import React, { createContext } from 'react';
import PropTypes from 'prop-types';
import { useContainerApartaments } from './container-theme';

const ApartamentsContext = createContext();

const ApartamentsProvider = ({ children, initialState = {} }) => {
  const contextValue = useContainerApartaments(initialState);

  return <ApartamentsContext.Provider value={contextValue}>{children}</ApartamentsContext.Provider>;
};

ApartamentsProvider.defaultProps = {
  initialState: {},
};

ApartamentsProvider.propTypes = {
  children: PropTypes.node.isRequired,
  initialState: PropTypes.shape({}),
};

ApartamentsContext.displayName = 'ApartamentsContext';

export { ApartamentsProvider, ApartamentsContext };
