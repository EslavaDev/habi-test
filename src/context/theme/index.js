import React, { createContext } from 'react';
import PropTypes from 'prop-types';
import { useContainerTheme } from './container-theme';

const ThemeContext = createContext();

const ThemeProvider = ({ children, initialState = {} }) => {
  const contextValue = useContainerTheme(initialState);

  return <ThemeContext.Provider value={contextValue}>{children}</ThemeContext.Provider>;
};

ThemeProvider.defaultProps = {
  initialState: {},
};

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
  initialState: PropTypes.shape({}),
};

ThemeContext.displayName = 'ThemeContext';

export { ThemeProvider, ThemeContext };
