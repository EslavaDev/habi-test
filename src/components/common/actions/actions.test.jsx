/* import { cleanup, fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Actions } from '.';
import { ApartamentsProvider } from '../../../context/apartaments';

let props = null;
let context = null;

const spy = jest.fn();

describe('Actions', () => {
  beforeEach(() => {
    spy.mockClear();
    props = {
      id: 'apt1',
    };
    context = {
      apartaments: null,
      apartamentsMap: {},
      apartamentsLike: {
        apt1: {},
      },
      dateFilter: null,
      locationFilter: null,
    };
  });

  afterEach(() => {
    cleanup();
  });

  test('Should render the component actions', () => {
    const { container } = render(<Actions {...props} />, { wrapper: ApartamentsProvider, initialState: context });
    expect(container).toMatchSnapshot();
  });

  test('Should render the correct fields', () => {
    const { getByTestId } = render(<Actions {...props} />, { wrapper: ApartamentsProvider, initialState: context });

    const action = getByTestId('action-heart2');

    fireEvent.click(action);

    expect(spy).toHaveBeenCalled();
  });
});
 */
