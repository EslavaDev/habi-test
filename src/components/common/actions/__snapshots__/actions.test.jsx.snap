// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions Should render the component actions 1`] = `
<div>
  <div
    class="col-6"
  >
    <svg
      aria-hidden="true"
      class="StyledIconBase-ea9ulj-0 fBxgqm"
      color="#dd3f5b"
      data-testid="action-heart1"
      fill="currentColor"
      focusable="false"
      height="20"
      viewBox="0 0 512 512"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z"
        fill="currentColor"
      />
    </svg>
  </div>
  <div
    class="col-6"
  >
    <svg
      aria-hidden="true"
      class="StyledIconBase-ea9ulj-0 fBxgqm"
      fill="currentColor"
      focusable="false"
      height="20"
      viewBox="0 0 448 512"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M268 416h24a12 12 0 0012-12V188a12 12 0 00-12-12h-24a12 12 0 00-12 12v216a12 12 0 0012 12zM432 80h-82.41l-34-56.7A48 48 0 00274.41 0H173.59a48 48 0 00-41.16 23.3L98.41 80H16A16 16 0 000 96v16a16 16 0 0016 16h16v336a48 48 0 0048 48h288a48 48 0 0048-48V128h16a16 16 0 0016-16V96a16 16 0 00-16-16zM171.84 50.91A6 6 0 01177 48h94a6 6 0 015.15 2.91L293.61 80H154.39zM368 464H80V128h288zm-212-48h24a12 12 0 0012-12V188a12 12 0 00-12-12h-24a12 12 0 00-12 12v216a12 12 0 0012 12z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
